### PR TITLE
Planning:use cuda to improve planning

### PR DESCRIPTION
**Pull request by muleisheng**

### CODE REVIEW

File: modules/planning/open_space/trajectory_smoother/distance_approach_ipopt_interface.cc [hunk 1]
```diff
@@ -2288,23 +2289,30 @@ bool DistanceApproachIPOPTInterface::eval_h(int n, const double* x, bool new_x,
...
+        }
     }
   } else {
     // return the values. This is a symmetric matrix, fill the lower left
     // triangle only
 
     obj_lam[0] = obj_factor;
-    for (int idx = 0; idx < m; idx++) obj_lam[1 + idx] = lambda[idx];
+    if (!data_transfer(&obj_lam[1], lambda, m)) {
+        for (int idx = 0; idx < m; idx++) obj_lam[1 + idx] = lambda[idx];
```
Comment (xiaoxq) [inline modules/planning/open_space/trajectory_smoother/distance_approach_ipopt_interface.cc]: Fix indents. Plz try "scripts/clang-format.sh" if you are not sure about styles.

File: modules/planning/open_space/trajectory_smoother/planning_block.cu [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "planning_block.h"
+#include <iostream>
+
+namespace apollo {
+namespace planning {
+bool InitialCuda() {
+    int dev = 0;
```
Comment (xiaoxq) [inline modules/planning/open_space/trajectory_smoother/planning_block.cu]: We also prefer the same coding style for .cu files though it's not checked by "apollo.sh lint".
Plz try "scripts/clang-format.sh /path/to/some.cu".

File: modules/planning/open_space/trajectory_smoother/planning_block.cu [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    cudaDeviceReset();
+    return true;
+}
+
+DATA_TRANSFER_INST(int);
+DATA_TRANSFER_INST(double);
+DATA_TRANSFER_INST(float);
+
+}
+}
```
Comment (xiaoxq) [inline modules/planning/open_space/trajectory_smoother/planning_block.cu]: Closing comment of // namespace.

**Comment by JasonZhou404** [comment]
You can run "python /apollo/modules/tools/open_space_visualization/distance_approach_visualizer.py". Compare the speed and result with and without your improvement

**Comment by Capri2014** [comment]
@muleisheng Could you also add the gflag for the protection as we discussed offline?

**Comment by Capri2014** [comment]
@JasonZhou404 The speed would not increase much, as this is just an initial kickoff trails without full implementation yet.

**Comment by deidaraho** [comment]
> @JasonZhou404 The speed would not increase much, as this is just an initial kickoff trails without full implementation yet.
It is a great start, is it possible to introduce a config to switch between implementations with and without cuda?

File: modules/planning/open_space/trajectory_smoother/distance_approach_ipopt_interface.cc (hunk 1 above)
Comment (muleisheng) [reply modules/planning/open_space/trajectory_smoother/distance_approach_ipopt_interface.cc]: Sorry for my fault. I fix this issue ASAP

**Comment by muleisheng** [comment]
> > @JasonZhou404 The speed would not increase much, as this is just an initial kickoff trails without full implementation yet.
> 
> @muleisheng It is a great start, is it possible to introduce a config to switch between implementations with and without cuda?

OK, I am modifying this pr and push it again

**Comment by muleisheng** [comment]
> @muleisheng Could you also add the gflag for the protection as we discussed offline?

I an using gflag and push it again today

**Comment by deidaraho** [comment]
@muleisheng Please fix the building conflicts.

**Pull request merged**